File: PE_0665/PE_0665.jl
# Michael Hunt
# 01-March -2020
# Proportionate Nim

# gradient_low=1.4777997752722001205


# 11541685709674 correct for M=10^7! in 4s


function mex(Set,minval)
    mexval = minval
    while Set[mexval+1] >=0
        mexval += 1
    end
    return mexval
end

function p665(M)

    losers=[(1,3),(2,6),(4,5),(7,10),(8,14),(9,17),(11,25),(12,28),(13,18),(15,35),(16,23),(19,31)]
    total=Int64(sum([k[1]+k[2] for k in losers]))
    nms=zeros(Int64,M).-1
    deltas=zeros(Bool,M)
    dhalfs=zeros(Bool,2*M)
    ddoubles=zeros(Bool,M)

    for i=1:length(losers)

        nms[losers[i][1]+1]=losers[i][1]
        nms[losers[i][2]+1]=losers[i][2]
        deltas[abs(losers[i][2]-losers[i][1])+1]=true
        dhalfs[abs(losers[i][1]-2*losers[i][2])+1]=true
        ddoubles[abs(2*losers[i][1]-losers[i][2])+1]=true
    end

    n=maximum(loser[1] for loser in losers)
    counthigh=5

    margin=200

    while 1>0

        n=mex(nms,n)
        nms[n+1]=n
        mcount=0

        gradient_low=1.4778
        m=max(1,floor(Int64,gradient_low*n)-margin)
        if n+m>M
            break
        end

        flag=false
        while flag==false
            mcount+=1
            if n+m>M || mcount>2*margin || m/n>1.56
                break
            end

            check=nms[m+1]>=0 ||
                + deltas[abs(m-n)+1]==true ||
                + dhalfs[abs(n-2*m)+1]==true ||
                + dhalfs[abs(m-2*n)+1]==true ||
                + ddoubles[abs(2*n-m)+1]==true
            if check==true
                m+=1
                continue
            end

            total+=n+m
            nms[m+1]=m
            deltas[abs(m-n)+1]=true
            dhalfs[abs(n-2*m)+1]=true
            ddoubles[abs(2*n-m)+1]=true
            flag=true
            break
        end

        if flag==true
            continue
        end

        counthigh+=1
        m=2*n+counthigh

        if n+m>M
            continue
        end

        total+=n+m
        nms[m+1]=m
        deltas[abs(m-n)+1]=true
        dhalfs[abs(n-2*m)+1]=true
        ddoubles[abs(2*n-m)+1]=true
    end
    println((M,total))
end


function test(M)

# s=Set(1:M)
# if a in s
#     i=1
# end
# s=Set()
s=zeros(M)
for a = 1:M
    # push!(s,a)
    s[a]=a
end

a=0
    for i=1:M
        for j=1:280
            if div(M,2) in s
                a=1
            end
        end
    end
end

File: PE_0665/PE_0665.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""

PE_0665

Proportionate Nim

Created on Thu Feb 20 16:30:12 2020

@author: mbh
"""

import matplotlib.pyplot as plt
import numpy as np

# find maximum excludant (mex) of a set
def mex(Set):
    mexval = 0
    while mexval in Set:
        mexval += 1
    return mexval


def grundyProp(n,m):
    
    ss=set()
    n,m=min((n,m)),max((n,m))
    for k in range(0,n):
        ss.add(tuple(sorted([k,m])))
    for k in range(0,m):
        ss.add(tuple(sorted([n,k])))
    for k in range(1,min(n,m)+1):
        ss.add(tuple(sorted([n-k,m-k])))
    for k in range(1,min(n,m//2)+1):
        ss.add(tuple(sorted([n-k,m-2*k])))
    for k in range(1,min(m,n//2)+1):
        ss.add(tuple(sorted([n-2*k,m-k])))
       

    
    for s in ss:
        print (s)
        
def p665(M):
    
    # total=0
    # count=0
    # ms=[]
    # ns=[]
    # for m in range(M+1):
    #     for n in range(min(m,M-m)):
    #         if grundyPropNim(n,m)==0:
    #             ms.append(m)
    #             ns.append(n)
    #             print(n,m)
    #             total+=n+m
    #             count+=1
    # print()
    # print(total,count)
    
    # print()
    
    
    # plt.plot(ms,ns,'bo')
    # plt.show()
    
    # total50=0
    # for i in range(len(ns)):
    #     if ns[i]<=50:
    #         total50+=ns[i]+ms[i]
            
    # print(total50)
    # print()
    
    count=0
    total=2193
    
    ah,bh=2.248793,0
    al,bl=1.478231 ,0
    
    dd=0.01
    print(int((M-bh)/(ah+1))+1)
    for n in range(50,int((M-bh)/(ah+1))+1):
        for m in range(int((1-dd)*(bh+ah*n)),int((1+dd)*(bh+ah*n))):
            if grundyPropNim(n,m)==0:
                print(n,m)
                total+=n+m
                count+=1
    print(int((M-bl)/(al+1))+1)
    for n in range(50,int((M-bl)/(al+1))+1):                  
        for m in range(int((1-dd)*(bl+al*n)),int((1+dd)*(bl+al*n))):
            if grundyPropNim(n,m)==0:
                print(n,m)
                total+=n+m
                count+=1
    print(total)

                
def printGrid(M):
    grid=np.zeros((M,M),dtype=np.int64)
    ns=np.zeros(M)
    ms=np.zeros(M)
    for n in range(M):
        for m in range(M):
            grid[n,m]=grundyPropNim(n,m)
            if grid[n,m]==0 and n<=m:
                print(n,m,m-n)
                ns[n]=n
                ms[m]=m
    # print(np.transpose(grid))
    # for n in range(M):
    #     for m in range(M):
    #         grid[n,m]=grundyWythoff(n,m)
    #         if grid[n,m]==0 and n<=m:
    #             print(n,m)
    # print(np.transpose(grid))         
                

def WythoffZeros(M):
    gr=(1+np.sqrt(5))/2
    for n in range(M):
        print(int(gr*n),int(gr**2*n))
       
        
def grundyPropNim(n,m,memo={}):
    if n==0 and m==0:
        return 0
    if n==0 and m==1:
        return 1
    if n==1 and m==0:
        return 1


    try:
        return memo[n,m]
    except KeyError:
        gset=set()
        for k in range(1,n+1):
            result=grundyPropNim(n-k,m,memo)
            gset.add(result)
            memo[n,m]=mex(gset)
        for k in range(1,m+1):
            result=grundyPropNim(n,m-k,memo)
            gset.add(result)
            memo[n,m]=mex(gset)

        nm_min=min(n,m)
        nm_diff=abs(n-m)
        for k in range(1,nm_min+1):
            result=grundyPropNim(nm_min-k,nm_min+nm_diff-k,memo)
            gset.add(result)
            memo[n,m]=mex(gset)
            
        for k in range(1,min(nm_min,(nm_min+nm_diff)//2)+1):
            result=grundyPropNim(nm_min-k,nm_min+nm_diff-2*k,memo)
            gset.add(result)
            memo[n,m]=mex(gset)
            
        for k in range(1,nm_min//2+1):
            result=grundyPropNim(nm_min-2*k,nm_min+nm_diff-k,memo)
            gset.add(result)
            memo[n,m]=mex(gset)

        return mex(gset)

def grundyWythoff(n,m,memo={}):

    if n==0 and m==0:
        return 0
    if n==0 and m==1:
        return 1
    if n==1 and m==0:
        return 1

    try:
        return memo[n,m]
    except KeyError:
        gset=set()
        for k in range(1,n+1):
            result=grundyWythoff(n-k,m,memo)
            gset.add(result)
            memo[n,m]=mex(gset)
        for k in range(1,m+1):
            result=grundyWythoff(n,m-k,memo)
            gset.add(result)
            memo[n,m]=mex(gset)

        nm_min=min(n,m)
        nm_diff=abs(n-m)
        for k in range(1,nm_min+1):
            result=grundyWythoff(nm_min-k,nm_min+nm_diff-k,memo)
            gset.add(result)
            memo[n,m]=mex(gset)

        return mex(gset)